loki/expression/expr_visitors.py

```diff
@@ -138,6 +138,17 @@ def visit_TypeDef(self, o, **kwargs):
         """
         return self._return(o, ())
 
+    def visit_VariableDeclaration(self, o, **kwargs):
+        expressions = ()
+        for v in o.symbols:
+            if v.type.initial is not None:
+                retrieved = self.retrieve(v.type.initial)
+                if retrieved:
+                    expressions += as_tuple(retrieved)
+        if expressions:
+            return self._return(o, expressions)
+        return super().visit(o.children, **kwargs)
```

Comment:
I'm not sure I'm following the logic here. The real purpose of this visitor method is to include the `initial` attribute in the lookup, no? However, the implementation here reads as if it has the side-effect of not searching the other expressions in the declaration anymore (e.g., a `dimensions`/`shape` attribute) if any `initial` attribute yielded a relevant return value.

Wouldn't something like this be an equivalent and more straightforward implementation?
```py
expressions = super().visit(o, **kwargs)
for v in o.symbols:
    if v.type.initial is not None:
        expressions += as_tuple(self.retrieve(v.type.initial))
return self._return(o, expressions)
```

Reply:
`expressions = super().visit(o, **kwargs)` gives me a problem with recursion ...